### ISSUE TITLE
How to save a global data when I'm developing a new lint.

### ISSUE THREAD

**Issue by callanchen** [rust-lang/rust-clippy#14170]
This is my ui test file:
```rust
fn foo1() {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();  // input is marked as tainted

    // This line will trigger a lint warning
    let file = std::fs::File::open(input).unwrap();
}

fn foo2() {
    let mut input = String::new();

    io::stdin().read_line(&mut input).unwrap();  // input is marked as tainted
    
    // This line is correct usage
    let file = std::fs::File::open(validate(&input)).unwrap();
}
```

I want to implement a new lint: variables that store user input, when used, should be validated.

So, I want to save the variables which are tainted by inputing.

This is my lint:

```rust
#[derive(Default)]
#[derive(Debug)]
struct TaintState {
    tainted_vars: HashSet<rustc_hir::HirId>,
}

declare_lint_pass!(ProtectAgainstFileNameInjection => [PROTECT_AGAINST_FILE_NAME_INJECTION]);

impl<'tcx> LateLintPass<'tcx> for ProtectAgainstFileNameInjection {
    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {

        let mut state = TaintState::default();
        // Check if it reads from standard input
        if let ExprKind::MethodCall(path, receiver, args, _) = &expr.kind {
            if path.ident.name.as_str() == "read_line" {
            // If it is a read_line call, mark the receiving variable as tainted
            state.tainted_vars.insert(receiver.hir_id);
            }
        }

        // !!!!!!!!!! This is the problem !!!!!!!!!!
        println!("{:?}", state); // Every time the lint is run, the state is reset to default

        // Check file operations
        if let ExprKind::MethodCall(path, _, args, _) = &expr.kind {
            if path.ident.name.as_str() == "open" {
            // Check if the first argument is a tainted path
            if let Some(path_expr) = args.first() {
                if state.tainted_vars.contains(&path_expr.hir_id) {
                // Check if it is wrapped by the validate function
                if !is_validated(path_expr) {
                    span_lint_and_help(
                    cx,
                    PROTECT_AGAINST_FILE_NAME_INJECTION,
                    expr.span,
                    "using untrusted input as file path",
                    None,
                    "consider using the validate() function to check the path",
                    );
                }
                }
            }
            }
        }
        }
    }

    // Check if the expression is wrapped by the validate function
    fn is_validated(expr: &Expr) -> bool {
        if let ExprKind::Call(func, _) = &expr.kind {
        if let ExprKind::Path(qpath) = &func.kind {
            match qpath {
            QPath::Resolved(_, path) => {
                // Check the last path segment
                if let Some(segment) = path.segments.last() {
                return segment.ident.name.as_str() == "validate";
                }
            }
            QPath::TypeRelative(_, segment) => {
                // For type-relative functions
                return segment.ident.name.as_str() == "validate";
            }
                _ => {}
            }
        }
    }
    false
}
```

I searched many information for a long time, but I can not find a way to solve my problem.

I'm the newbie and this is my first time to develop a new lint.

If anyone could give me any advice, I'd really appreciate it.

**Comment by y21** [comment]
Instead of using the `declare_lint_pass!` macro, you can create a struct that stores your `TaintState` and use `impl_lint_pass!` instead. Then instead of creating a new `TaintState` every expression you access it via `self.taint_state`.

(Also, note that the `receiver.hir_id` that you're inserting there is not the HirId of the variable. The receiver is the `io::stdin()` expression. You probably meant to use the first argument instead of the receiver. And even then, the HirId of a variable path expression is not the same as the HirId of the actual variable. You can use [`clippy_utils::path_to_local`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.path_to_local.html) for that)

**Comment by callanchen** [comment]
Thank you very much.
What I want to ask you is that I have just finished reading Clippy Book, but there are still some things I don’t know.
Where should I learn more detailed information?

**Comment by y21** [comment]
What do you want to learn more about? I think the clippy book is the most detailed thing we have for clippy things specifically. I guess there's the [rustc dev guide](https://rustc-dev-guide.rust-lang.org) if you want to learn more about rustc internals. You can also always ask your questions on [zulip](https://rust-lang.zulipchat.com) in the clippy channel

**Comment by callanchen** [comment]
Thank you for your help, I have no further questions.
I closed the issue.
